### ISSUE TITLE
Implement BitXorAssign trait for Hash

### ISSUE THREAD

**Issue by Eitu33** [BLAKE3-team/BLAKE3#262]
There is currently no way to XOR two `blake3::Hash` types without making extra allocations, because it is not possible to iterate mutably over a `blake3::Hash`. It is possible to work around it but I think it is a common enough use case for it to be implemented in the crate. I could take care of this if you believe it is worthwhile.

**Comment by oconnor663** [comment]
Is the extra allocation you're talking about the 32 bytes on the stack that you'd expect from converting the `Hash` into a `[u8; 32]`? Are you sure that's actually a measurable cost in your use case? It might be interesting to look at how Rustc completely optimizes away that cost on x86-64:

```rust
pub fn xor_hashes(this: &mut blake3::Hash, other: &blake3::Hash) {
    let mut bytes: [u8; 32] = (*this).into();
    // Note that we could use a .zip() iterator here to guarantee no bounds checks for
    // the array accesses, but Rustc is smart enough that we don't need to bother.
    for i in 0..32 {
        bytes[i] ^= other.as_bytes()[i];
    }
    *this = bytes.into();
}
```

`cargo asm` shows me this assembly output on my machine (I've added comments):

```asm
; Copy the left half of this into a 128-bit SSE register.
movups  xmm0, xmmword, ptr, [rdi]
; Copy the right half of this into a 128-bit SSE register.
movups  xmm1, xmmword, ptr, [rdi, +, 16]
; Copy the left half of other into a 128-bit SSE register.
movups  xmm2, xmmword, ptr, [rsi]
; XOR the left halves of this and other, overwriting the third register.
xorps   xmm2, xmm0
; Copy the right half of other into the first register.
movups  xmm0, xmmword, ptr, [rsi, +, 16]
; XOR the right halves of this and other, overwriting the second register.
xorps   xmm0, xmm1
; Write the first XOR result to the left half of this.
movups  xmmword, ptr, [rdi], xmm2
; Write the second XOR result to the right half of this.
movups  xmmword, ptr, [rdi, +, 16], xmm0
ret
```

What we see in this case is that even the stack allocation is gone, and all the loading and xor'ing and storing is done with registers. If we set `RUSTFLAGS="-C target-cpu=native" on a reasonably recent CPU, the optimizer will get this down to a *single register* :)

```asm
vmovups ymm0, ymmword, ptr, [rdi]
vxorps  ymm0, ymm0, ymmword, ptr, [rsi]
vmovups ymmword, ptr, [rdi], ymm0
vzeroupper
ret
```

So my instinct here is that modern optimizing compilers are so smart that there's no actual performance cost here in either space or time. If that's right, then I lean towards not adding an XOR operator, to keep things simple and to avoid any possible confusion about what such an operator is doing.

**Comment by damip** [comment]
Thank you for the analysis @oconnor663 ! Out of curiosity, do you observe the same optimization for `release` but non-`target-cpu=native` compilation (eg. when distributing binaries generated by CD) ? If so, I agree with you 100%. If not, then I see a use case for this operator.

**Comment by oconnor663** [comment]
Yes I expect that the `--release` mode defaults on x86-64 will result in code like the first asm output above. (The SSE extensions predate x86-64, and I think all the common compilers will assume that SSE is available when targeting 64-bit.)
